pkg/driver/controller_modify_volume.go

```diff
@@ -107,55 +113,106 @@ func mergeModifyVolumeRequest(input modifyVolumeRequest, existing modifyVolumeRe
+		err := executeModifyTagsRequest(volumeID, req, c)
 		if err != nil {
-			// Kubernetes sidecars treats "Invalid Argument" errors as infeasible and retries less aggressively
-			if errors.Is(err, cloud.ErrInvalidArgument) {
-				return 0, status.Errorf(codes.InvalidArgument, "Could not modify volume (invalid argument) %q: %v", volumeID, err)
+			return 0, err
+		}
+		if (req.modifyDiskOptions.IOPS != 0) || (req.modifyDiskOptions.Throughput != 0) || (req.modifyDiskOptions.VolumeType != "") || (req.newSize != 0) {
+			actualSizeGiB, err := c.ResizeOrModifyDisk(ctx, volumeID, req.newSize, &req.modifyDiskOptions)
+			if err != nil {
+				if errors.Is(err, cloud.ErrInvalidArgument) {
+					return 0, status.Errorf(codes.InvalidArgument, "Could not modify volume (invalid argument) %q: %v", volumeID, err)
+				}
+				return 0, status.Errorf(codes.Internal, "Could not modify volume %q: %v", volumeID, err)
+			} else {
+				return actualSizeGiB, nil
 			}
-			return 0, status.Errorf(codes.Internal, "Could not modify volume %q: %v", volumeID, err)
-		} else {
-			return actualSizeGiB, nil
 		}
+		//Any error should be caught in one of the conditional statements above, returning 0 here is correct if it doesn't go into the second conditional
+		return 0, nil
 	}
 }
 
-func parseModifyVolumeParameters(params map[string]string) (*cloud.ModifyDiskOptions, error) {
-	options := cloud.ModifyDiskOptions{}
-
+func parseModifyVolumeParameters(params map[string]string) (*modifyVolumeRequest, error) {
+	options := modifyVolumeRequest{
+		modifyTagsOptions: &cloud.ModifyTagsOptions{
+			TagsToAdd:    make(map[string]string),
+			TagsToDelete: make([]string, 0),
+		},
+	}
 	for key, value := range params {
 		switch key {
 		case ModificationKeyIOPS:
 			iops, err := strconv.Atoi(value)
 			if err != nil {
 				return nil, status.Errorf(codes.InvalidArgument, "Could not parse IOPS: %q", value)
 			}
-			options.IOPS = int32(iops)
+			options.modifyDiskOptions.IOPS = int32(iops)
 		case ModificationKeyThroughput:
 			throughput, err := strconv.Atoi(value)
 			if err != nil {
 				return nil, status.Errorf(codes.InvalidArgument, "Could not parse throughput: %q", value)
 			}
-			options.Throughput = int32(throughput)
+			options.modifyDiskOptions.Throughput = int32(throughput)
 		case DeprecatedModificationKeyVolumeType:
 			if _, ok := params[ModificationKeyVolumeType]; ok {
 				klog.Infof("Ignoring deprecated key `volumeType` because preferred key `type` is present")
 				continue
 			}
 			klog.InfoS("Key `volumeType` is deprecated, please use `type` instead")
-			options.VolumeType = value
+			options.modifyDiskOptions.VolumeType = value
 		case ModificationKeyVolumeType:
-			options.VolumeType = value
+			options.modifyDiskOptions.VolumeType = value
+		default:
+			if strings.HasPrefix(key, ModificationAddTag) {
+				st := strings.SplitN(value, "=", 2)
+				options.modifyTagsOptions.TagsToAdd[st[0]] = st[1]
```

Comment:
Add a check here for if `len(st) < 2` (like if the user provides an invalid input without an equal sign), and add a test for that case too